core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -233,70 +245,83 @@ private[transaction] class TransactionMetadata(val producerId: Long,
           } else {
-            txnStartTimestamp = newMetadata.txnStartTimestamp
-            addPartitions(newMetadata.topicPartitions)
+            txnStartTimestamp = transitMetadata.txnStartTimestamp
+            addPartitions(transitMetadata.topicPartitions)
           }
 
         case PrepareAbort | PrepareCommit => // from endTxn
-          if (producerEpoch != newMetadata.producerEpoch ||
-            !topicPartitions.toSet.equals(newMetadata.topicPartitions) ||
-            txnTimeoutMs != newMetadata.txnTimeoutMs ||
-            txnStartTimestamp != newMetadata.txnStartTimestamp) {
+          if (producerEpoch != transitMetadata.producerEpoch ||
+            !topicPartitions.toSet.equals(transitMetadata.topicPartitions) ||
+            txnTimeoutMs != transitMetadata.txnTimeoutMs ||
+            txnStartTimestamp != transitMetadata.txnStartTimestamp) {
 
             throw new IllegalStateException("Completing transaction state transition failed due to unexpected metadata")
           }
 
         case CompleteAbort | CompleteCommit => // from write markers
-          if (producerEpoch != newMetadata.producerEpoch ||
-            txnTimeoutMs != newMetadata.txnTimeoutMs ||
-            newMetadata.txnStartTimestamp == -1) {
+          if (producerEpoch != transitMetadata.producerEpoch ||
+            txnTimeoutMs != transitMetadata.txnTimeoutMs ||
+            transitMetadata.txnStartTimestamp == -1) {
 
             throw new IllegalStateException("Completing transaction state transition failed due to unexpected metadata")
           } else {
-            txnStartTimestamp = newMetadata.txnStartTimestamp
+            txnStartTimestamp = transitMetadata.txnStartTimestamp
             topicPartitions.clear()
           }
       }
 
-      txnLastUpdateTimestamp = newMetadata.txnLastUpdateTimestamp
+      debug(s"TransactionalId $transactionalId complete transition from $state to $transitMetadata")
+      txnLastUpdateTimestamp = transitMetadata.txnLastUpdateTimestamp
       pendingState = None
       state = toState
     }
   }
 
   def pendingTransitionInProgress: Boolean = pendingState.isDefined
 
-  override def toString = s"TransactionMetadata($pendingState, $producerId, $producerEpoch, $txnTimeoutMs, $state, $topicPartitions, $txnStartTimestamp, $txnLastUpdateTimestamp)"
+  override def toString = {
+    "TransactionMetadata(" +
+      s"transactionalId=$transactionalId, " +
+      s"producerId=$producerId, " +
+      s"producerEpoch=$producerEpoch, " +
+      s"txnTimeoutMs=$txnTimeoutMs, " +
+      s"state=$state, " +
+      s"pendingState=$pendingState, " +
+      s"topicPartitions=$topicPartitions, " +
+      s"txnStartTimestamp=$txnStartTimestamp, " +
+      s"txnLastUpdateTimestamp=$txnLastUpdateTimestamp)"
+  }
 
   override def equals(that: Any): Boolean = that match {
     case other: TransactionMetadata =>
+      transactionalId == other.transactionalId &&
```

Comment:
This field is added just for logging? In that case we could just add the logging in the caller of `completeTransitionTo` and `prepareTransitionTo`.